### PR TITLE
Improve the crypto doc

### PR DESCRIPTION
**Pull request by tarakby**
## Description

Add more details to the crypto contract and native crypto features :
 - move the public key section to `crypto.md` as `PublicKey` is a cryptographic structure rather than an account structure. 
 - add details about signature verification.
 - add details about hashing with and without tags.

### CODE REVIEW

File: docs/language/crypto.md [hunk 1]
```diff
@@ -39,20 +45,147 @@ pub enum SignatureAlgorithm: UInt8 {
...
+publicKey.signatureAlgorithm = SignatureAlgorithm.ECDSA_secp256k1   // Not allowed
+publicKey.publicKey = []                                            // Not allowed
+
+publicKey.publicKey[2] = 4      // No effect
+```
+
+The validity of a public key can be checked using the `isValid` field.
+Verifications performed with an invalid public key (using `verify()` method) will always fail.
+
+#### Signature verification
```
Comment (turbolent) [inline docs/language/crypto.md:100]:
```suggestion
#### Signature verification

A signature can be verified using the `verify` function of the `PublicKey`:

```

File: docs/language/crypto.md [hunk 2]
```diff
@@ -39,20 +45,147 @@ pub enum SignatureAlgorithm: UInt8 {
...
+
+```cadence
+// Hash the data using the given hashing algorithm and returns the hashed data.
+pub fun hash(_ data: [UInt8], algorithm: HashAlgorithm): [UInt8]
+
+// Hash the data using the given hashing algorithm and the tag. Returns the hashed data.
+pub fun hashWithTag(_ data: [UInt8], tag: string, algorithm: HashAlgorithm): [UInt8]
+```
+
+🚧 Status: `KMAC128_BLS_BLS12_381` is not supported yet.
```
Comment (turbolent) [inline docs/language/crypto.md:154]:
```suggestion
<Callout type="info">

🚧 Status: `KMAC128_BLS_BLS12_381` is not supported yet.

</Callout>
```

File: docs/language/crypto.md [hunk 3]
```diff
@@ -39,20 +45,147 @@ pub enum SignatureAlgorithm: UInt8 {
...
+
+- `hash` hashes the input data using the input hashing algorithm. 
+    - If `KMAC128_BLS_BLS12_381` is used, data is hashed using the standard mac `KMAC128(customizer, key, data, length)` with the following inputs:
+        - `customizer` is the UTF-8 encoding of "H2C" 
+        - `key` is the UTF-8 encoding of `"APP_BLS_SIG_BLS12381G1_XOF:KMAC128_SSWU_RO_POP_"`
+        - `data` is the input data to hash
+        - `length` is 1024 bytes
+
+🚧 Status: `hashWithTag` is not supported yet.
```
Comment (turbolent) [inline docs/language/crypto.md:163]:
```suggestion

<Callout type="info">

🚧 Status: `hashWithTag` is not supported yet.

</Callout>
```

File: docs/language/crypto.md [hunk 4]
```diff
@@ -39,20 +45,147 @@ pub enum SignatureAlgorithm: UInt8 {
...
+
+// Hash the data using the given hashing algorithm and the tag. Returns the hashed data.
+pub fun hashWithTag(_ data: [UInt8], tag: string, algorithm: HashAlgorithm): [UInt8]
+```
+
+🚧 Status: `KMAC128_BLS_BLS12_381` is not supported yet.
+
+- `hash` hashes the input data using the input hashing algorithm. 
+    - If `KMAC128_BLS_BLS12_381` is used, data is hashed using the standard mac `KMAC128(customizer, key, data, length)` with the following inputs:
+        - `customizer` is the UTF-8 encoding of "H2C"
```
Comment (turbolent) [inline docs/language/crypto.md:158]:
```suggestion
        - `customizer` is the UTF-8 encoding of `"H2C"`
```

File: docs/language/crypto.md [hunk 5]
```diff
@@ -39,20 +45,147 @@ pub enum SignatureAlgorithm: UInt8 {
...
+        - `key` is the UTF-8 encoding of `"APP_" || tag || "BLS_SIG_BLS12381G1_XOF:KMAC128_SSWU_RO_POP_"`
+        - `data` is the input data to hash
+        - `length` is 1024 bytes
+
+
+For example, to compute a SHA3-256 hash:
+
+```cadence
+ let data = "random_data"
+ let digest = hash(data, SHA3_256)
```
Comment (turbolent) [inline docs/language/crypto.md:178]:
```suggestion
let data: [UInt8] = [1, 2, 3]
let digest = Crypto.hash(data, algorithm: HashAlgorithm.SHA3_256)
```

**Review by turbolent — COMMENTED** [review]
Thank you for adding details and the current status!

Looks good, left some suggestions

File: docs/language/crypto.md (hunk 5 above)
Comment (tarakby) [reply docs/language/crypto.md:178]: Thanks for fixing this!

File: docs/language/crypto.mdx [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+```
+
+The validity of a public key can be checked using the `isValid` field.
+Verifications performed with an invalid public key (using `verify()` method) will always fail.
+
+#### Signature verification
+
+A signature can be verified using the `verify` function of the `PublicKey`:
+
+```cadence
```
Comment (tarakby) [inline docs/language/crypto.mdx:128]: @turbolent can you please have a look a this code?

File: docs/language/crypto.mdx [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+let signature = "108EF718F153CFDC516D8040ABF2C8CC7AECF37C6F6EF357C31DFE1F7AC79C9D0145D1A2F08A48F1A2489A84C725D6A7AB3E842D9DC5F8FE8E659FFF5982310D"
+let message : [UInt8] = [1, 2, 3]
+
+let isValid = pk.verify(   
+       signature: signature,
+        signedData: message,
+        domainSeparationTag: "",
+        hashAlgorithm: SHA2_256
+    ) // isValid should be false
```
Comment (turbolent) [inline docs/language/crypto.mdx:142]:
```suggestion
let signature = "108EF718F153CFDC516D8040ABF2C8CC7AECF37C6F6EF357C31DFE1F7AC79C9D0145D1A2F08A48F1A2489A84C725D6A7AB3E842D9DC5F8FE8E659FFF5982310D".decodeHex()
let message : [UInt8] = [1, 2, 3]

let isValid = pk.verify(   
    signature: signature,
    signedData: message,
    domainSeparationTag: "",
    hashAlgorithm: HashAlgorithm.SHA2_256
)
// `isValid` is false
```

Comment (tarakby) [reply docs/language/crypto.mdx]: Thank you! 🙏

**Comment by turbolent** [comment]
@tarakby There are a few things that I suggested that weren't applied.

The easiest way to apply the suggestions is to use the GitHub feature it has built-in support for committing individual suggestions and even committing multiple suggestions in one commit as a batch.

File: docs/language/crypto.mdx [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+---
+
+### Hashing Algorithms
+The built-in enum `HashAlgorithm` provides the set of hashing algorithms that
+are supported by the language natively.
+
+<Callout type="info">
+
+🚧 Status: `KMAC128_BLS_BLS12_381` is not supported yet.
+
```
Comment (turbolent) [inline docs/language/crypto.mdx:12]:
```suggestion

</Callout>
```

File: docs/language/crypto.mdx [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+ The public key is a curve point `(X,Y)` where `X` and `Y` are two prime field elements. 
+ The raw key is represented as `bytes(X) || bytes(Y)`, where `||` is the concatenation operation,
+  and `bytes()` is the bytes big-endian encoding of the input integer padded by zeros to the byte-length of the field prime.
+  The raw public key is 64-bytes long.
+
+ - `BLS_BLS_12_381` : 
+ The public key is a G_2 (curve over extension field) element. 
+ The encoding follows the compressed serialization defined in the
+ [IETF draft-irtf-cfrg-pairing-friendly-curves-08](https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-08.html#name-point-serialization-procedu).
+ A public key is 96-bytes long.
```
Comment (turbolent) [inline docs/language/crypto.mdx:109]:
```suggestion
- `ECDSA_P256` and `ECDSA_secp256k1`:
  The public key is a curve point `(X,Y)` where `X` and `Y` are two prime field elements. 
  The raw key is represented as `bytes(X) || bytes(Y)`, where `||` is the concatenation operation,
  and `bytes()` is the bytes big-endian encoding of the input integer padded by zeros to the byte-length of the field prime.
  The raw public key is 64-bytes long.

- `BLS_BLS_12_381`:
  The public key is a G_2 (curve over extension field) element. 
  The encoding follows the compressed serialization defined in the
  [IETF draft-irtf-cfrg-pairing-friendly-curves-08](https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-08.html#name-point-serialization-procedu).
  A public key is 96-bytes long.
```

File: docs/language/crypto.mdx [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+let signature = "108EF718F153CFDC516D8040ABF2C8CC7AECF37C6F6EF357C31DFE1F7AC79C9D0145D1A2F08A48F1A2489A84C725D6A7AB3E842D9DC5F8FE8E659FFF5982310D".decodeHex()
+let message : [UInt8] = [1, 2, 3]
+
+let isValid = pk.verify(   
+       signature: signature,
+        signedData: message,
+        domainSeparationTag: "",
+        hashAlgorithm: HashAlgorithm.SHA2_256
+    ) 
+    // isValid should be false
```
Comment (turbolent) [inline docs/language/crypto.mdx:143]:
```suggestion
let isValid = pk.verify(   
    signature: signature,
    signedData: message,
    domainSeparationTag: "",
    hashAlgorithm: HashAlgorithm.SHA2_256
) 
// `isValid` is false
```

File: docs/language/crypto.mdx [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+   Only `SHA2_256` or `SHA3_256` are accepted.
+
+- BLS (`BLS_BLS_12_381`): 
+   - `signature` is a G_1 (curve over prime field) point. 
+   The encoding follows the compressed serialization defined in the [IETF draft-irtf-cfrg-pairing-friendly-curves-08](https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-08.html#name-point-serialization-procedu).
+   A signature is 48-bytes long.
+   - `signedData` is the arbitrary message to verify the signature against.
+   - `domainSeparationTag` is the domain tag used for signing. All tags are accepted. 
+   - `hashAlgorithm` is the algorithm used to hash the message along with the given tag (check `hashWithTag` function for more details). 
+   Only `KMAC128_BLS_BLS12_381` is accepted.
```
Comment (turbolent) [inline docs/language/crypto.mdx:171]:
```suggestion
   - `signature` is the couple `(r,s)`. It is represented as `bytes(r) || bytes(s)`, where `||` is the concatenation operation, 
     and `bytes()` is the bytes big-endian encoding of the input integer padded by zeros to the byte-length of the curve order. 
     The signature is 64 bytes-long for both curves. 
   - `signedData` is the arbitrary message to verify the signature against.
   - `domainSeparationTag` is the domain tag used for signing. The interface only accepts the the user tag, 
   the UTF-8 encoding of `"FLOW-V0.0-user"` padded to 32 bytes.
   - `hashAlgorithm` is the algorithm used to hash the message along with the given tag (check `hashWithTag` function for more details). 
     Only `SHA2_256` or `SHA3_256` are accepted.

- BLS (`BLS_BLS_12_381`): 
   - `signature` is a G_1 (curve over prime field) point. 
     The encoding follows the compressed serialization defined in the [IETF draft-irtf-cfrg-pairing-friendly-curves-08](https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-08.html#name-point-serialization-procedu).
     A signature is 48-bytes long.
   - `signedData` is the arbitrary message to verify the signature against.
   - `domainSeparationTag` is the domain tag used for signing. All tags are accepted. 
   - `hashAlgorithm` is the algorithm used to hash the message along with the given tag (check `hashWithTag` function for more details). 
     Only `KMAC128_BLS_BLS12_381` is accepted.
```

File: docs/language/crypto.mdx [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+For example, to compute a SHA3-256 hash:
+
+```cadence
+let data: [UInt8] = [1, 2, 3]
+let digest = Crypto.hash(data, algorithm: HashAlgorithm.SHA3_256)
+```
+
+The crypto contract also allows creating key lists to be used for a simple multi-signature 
+verification. 
+For example, to verify two signatures with equal weights for some signed data:
```
Comment (turbolent) [inline docs/language/crypto.mdx:229]: Just realized that this code sample showing how to use key lists to do multi-sig verification is now under the Hashing subsection, add a subsection here:

```suggestion

### Key Lists

The crypto contract also allows creating key lists to be used for multi-signature verification.
For example, to verify two signatures with equal weights for some signed data:
```

**Comment by tarakby** [comment]
@turbolent sorry my bad! I missed the little spaces you suggested, I thought they were optional. Thanks again for the thorough review!

**Comment by turbolent** [comment]
15c9251ac updated the docstring in the docs, so I also updated it in the code in 08ea84c

**Review by turbolent — APPROVED** [review]

**Pull request merged**